URI Fragments Cached After 301 Redirect

stwrt opened the pull request:
#### 205444531ae8c5213c84b0b63cd0d6ec51cf7735
<pre>
URI Fragments Cached After 301 Redirect
<a href="https://bugs.webkit.org/show_bug.cgi?id=255924">https://bugs.webkit.org/show_bug.cgi?id=255924</a>
rdar://problem/108535944

Reviewed by Youenn Fablet.

We end up reusing the fragments when performing a redirect and as such end up serving old parameters.
For now we will just disable caching redirects with fragments. This is a temporary fix; a more proper
solution of handling fragment caching in the future is required.

* LayoutTests/http/tests/links/redirect-301-clear-memory-cache-expected.txt: Added.
* LayoutTests/http/tests/links/redirect-301-clear-memory-cache.html: Added.
* LayoutTests/http/tests/links/redirect-301-no-inital-fragments-expected.txt: Added.
* LayoutTests/http/tests/links/redirect-301-no-inital-fragments.html: Added.
* LayoutTests/http/tests/links/redirect-301-with-fragments-expected.txt: Added.
* LayoutTests/http/tests/links/redirect-301-with-fragments.html: Added.
* LayoutTests/http/tests/links/resources/redirect-helper.pl: Added.
* LayoutTests/http/tests/links/resources/redirect-target.html: Added.
* Source/WebCore/loader/cache/CachedResource.cpp:
(WebCore::CachedResource::redirectReceived):
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::determineRevalidationPolicy const):
* Source/WebCore/loader/cache/CachedResourceRequest.h:
(WebCore::CachedResourceRequest::hasFragmentIdentifier const):
* Source/WebKit/NetworkProcess/cache/NetworkCache.cpp:
(WebKit::NetworkCache::makeUseDecision):
(WebKit::NetworkCache::makeStoreDecision):
* Source/WebKit/NetworkProcess/cache/NetworkCache.h:

Canonical link: <a href="https://commits.webkit.org/265906@main">https://commits.webkit.org/265906@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0293aa532ba4ac7d238287e894aace810e844200

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12133 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12480 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12782 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13879 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/11708 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12164 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14895 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12494 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14395 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12297 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13104 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10274 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14295 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10390 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11015 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/18131 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11471 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11174 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14348 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/11663 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9613 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10879 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/10890 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3002 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15205 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/11516 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/platform/network/ResourceRequestBase.h:
```diff
@@ -264,6 +264,8 @@ class ResourceRequestBase {
     bool useNetworkConnectionIntegrity() const { return m_requestData.m_useNetworkConnectionIntegrity; }
     WEBCORE_EXPORT void setUseNetworkConnectionIntegrity(bool);
 
+    bool urlContainsFragmentIdentifier() const { return m_requestData.m_firstPartyForCookies.hasFragmentIdentifier(); }
```
using `m_firstPartyForCookies` isn't right here. Call sites should use `request.url().hasFragmentIdentifier()`.

stwrt replied:
Removed this. Thanks was not sure how to access it otherwise.

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCache.cpp:
```diff
@@ -284,6 +284,12 @@ static StoreDecision makeStoreDecision(const WebCore::ResourceRequest& originalR
+    // FIXME: Our caching does not take into account fragments properly.
+    // If we cache localhost/#key=foo we will return the same parameters key=foo
+    // on the next visit to the domain even if the parameters have been updated.
+    if ((response.isRedirection() || response.isRedirected()) && originalRequest.urlContainsFragmentIdentifier())
```
Not sure why we need `response.isRedirection()`

stwrt replied:
This was handling loading the redirected page.
This might not be needed.

stwrt replied:
Was able to remove. Not needed.

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCache.cpp:
```diff
@@ -284,6 +284,12 @@ static StoreDecision makeStoreDecision(const WebCore::ResourceRequest& originalR
             return StoreDecision::NoDueToHTTPStatusCode;
     }
 
+    // FIXME: Our caching does not take into account fragments properly.
```
I don't think we need a FIXME comment.

stwrt replied:
Removed

stwrt commented on Source/WebCore/loader/cache/CachedResource.cpp:
```diff
@@ -454,6 +454,9 @@ void CachedResource::redirectReceived(ResourceRequest&& request, const ResourceR
 {
     CACHEDRESOURCE_RELEASE_LOG("redirectReceived:");
 
+    if(request.url().hasFragmentIdentifier())
```
This has a bug in the case of `localhost:10000` then loading `localhost:10000/#key=foo`.

stwrt replied:
Fixed.

cdumez commented on LayoutTests/imported/w3c/web-platform-tests/html/links/redirect-301-with-fragments.html:
```diff
@@ -0,0 +1,14 @@
+</head>
+<body>
+    <script>
+        
```
nit: extra line

cdumez replied:
Also, we need 2 tests, one to cover the disk cache change and the other to cover the memory cache one.

cdumez commented on Source/WebCore/loader/cache/CachedResource.cpp:
```diff
@@ -454,6 +454,9 @@ void CachedResource::redirectReceived(ResourceRequest&& request, const ResourceR
 {
     CACHEDRESOURCE_RELEASE_LOG("redirectReceived:");
 
+    if (request.url().hasFragmentIdentifier())
```
This needs a comment to explain why.

cdumez commented on Source/WebCore/loader/cache/MemoryCache.cpp:
```diff
@@ -118,6 +118,9 @@ bool MemoryCache::add(CachedResource& resource)
     if (resource.resourceRequest().httpMethod() != "GET"_s)
         return false;
 
+    if (resource.response().isRedirection() && resource.resourceRequest().url().hasFragmentIdentifier())
```
This needs a comment to explain why.

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCache.cpp:
```diff
@@ -284,6 +284,12 @@ static StoreDecision makeStoreDecision(const WebCore::ResourceRequest& originalR
             return StoreDecision::NoDueToHTTPStatusCode;
     }
 
+    // Our caching does not take into account fragments properly.
```
It does, in general, because we don't normally care about fragments. This is only an issue with redirects and the comment should be clearer.

cdumez commented on Source/WebKit/NetworkProcess/cache/NetworkCache.cpp:
```diff
@@ -284,6 +284,12 @@ static StoreDecision makeStoreDecision(const WebCore::ResourceRequest& originalR
+    // If we cache localhost/#key=foo we will return the same parameters key=foo
+    // on the next visit to the domain even if the parameters have been updated.
+    if (response.isRedirection() && originalRequest.url().hasFragmentIdentifier())
+        return StoreDecision::NoDueToRedirectContainingFragments;
```
The redirect does not necessarily contain a fragment, it is the original request that has a fragment.

cdumez reviewed (commented):
Code change looks good overall but this needs better code comments and better testing.

cdumez reviewed (changes requested):
Needs better test coverage.

cdumez commented on LayoutTests/http/wpt/html/links/redirect-301-with-fragments-test-network-cache.html:
```diff
@@ -0,0 +1,17 @@
+    <script>
+        /* Clear the memory cache and disable it */
+        internals.clearMemoryCache();
+        internals.pruneMemoryCacheToSize(0);
```
I would think this is equivalent to clearing. How does this disable the memory cache?

cdumez replied:
As I thought, this test doesn't cover the disk cache fix. I have verified by running the test locally without this part of the fix. Please address.

stwrt replied:
Will take a look at this

cdumez commented on LayoutTests/http/wpt/html/links/redirect-301-with-fragments-test-network-cache.html:
```diff
@@ -0,0 +1,17 @@
+        internals.clearMemoryCache();
+        internals.pruneMemoryCacheToSize(0);
+
+        window.location.href = '/html/links/301-response.py#key=foo'
```
This file needs to be in a resources/ folder or our infrastructure thinks it is a standalone test.

stwrt replied:
Yeah, I noticed that too last night. Going to fix it today

youennf commented on Source/WebCore/loader/cache/CachedResourceLoader.cpp:
```diff
@@ -1494,6 +1494,9 @@ CachedResourceLoader::RevalidationPolicy CachedResourceLoader::determineRevalida
     }
 
+    if (cachedResourceRequest.containsFragmentIdentifier() && existingResource->hasRedirections())
+        return Load;
```
We should be able to reuse these resources.
Could you file a bug about this and add a FIXME here?

stwrt replied:
Sure, I'll add a FIXME

youennf commented on Source/WebCore/loader/cache/CachedResourceRequest.h:
```diff
@@ -106,6 +106,7 @@ class CachedResourceRequest {
     const SecurityOrigin* origin() const { return m_origin.get(); }
     SecurityOrigin* origin() { return m_origin.get(); }
 
+    bool containsFragmentIdentifier() { return m_fragmentIdentifier.length() > 0; }
```
```suggestion
    bool containsFragmentIdentifier() const { return !m_fragmentIdentifier.isEmpty(); }
```

stwrt replied:
Thanks. Fixed

youennf commented on Source/WebKit/NetworkProcess/cache/NetworkCache.cpp:
```diff
@@ -233,6 +233,11 @@ static UseDecision makeUseDecision(NetworkProcess& networkProcess, PAL::SessionI
+    // We could have cached a redirect without a fragment and now may have
+    // a fragment in the URL.
+    if (request.url().hasFragmentIdentifier() && entry.redirectRequest())
+        return UseDecision::NoDueToRequestContainingFragments;
```
I thought that we were removing the fragment identifier before going to the network process. Can you check this?

stwrt replied:
Doesn't look like it.

```
(lldb) p request.url()
(const WTF::URL) { http://localhost:10000/#key=foo }
```

stwrt replied:
Added a breakpoint right at this line and dumped out the url output.

youennf replied:
ok

youennf commented on Source/WebKit/NetworkProcess/cache/NetworkCache.cpp:
```diff
@@ -284,6 +289,12 @@ static StoreDecision makeStoreDecision(const WebCore::ResourceRequest& originalR
+    // Our caching does not properly handle the case where we perform a redirect with fragments.
+    // If we cache localhost/#key=foo we will return the same parameters key=foo
+    // on the next visit to the domain even if the parameters have been updated.
+    if ((response.isRedirection() || response.isRedirected()) && originalRequest.url().hasFragmentIdentifier())
```
It might be good to clarify the message to something like:
```
// FIXME: We are not correctly computing the redirected request URL in case original request
// has a fragment identifier and response location URL does not have one. Let's not store it for now.
```

youennf commented on LayoutTests/imported/w3c/web-platform-tests/html/links/301-response.py:
```diff
@@ -0,0 +1,27 @@
+def main(request, response):
+    response.status = 200
```
Why is this script named 301-response.py given it returns a 200 response?
It might be good to add tests for 301 and others redirection HTTP codes.

It might also be good to add tests for redirection URLS with location URLs having frag IDs if we do not cover this case.

youennf commented on LayoutTests/imported/w3c/web-platform-tests/html/links/301-response.py:
```diff
@@ -0,0 +1,27 @@
+        }
+
+        if (window.testRunner)
+            testRunner.notifyDone();
```
We cannot have testRunner in WPT tests, this might not be needed anyway given we are using testharness.js.
Please move them to http/wpt for now, or please upstream these changes to WPT.

stwrt replied:
Thanks. Removed.

youennf commented on LayoutTests/http/tests/links/resources/redirect-helper.pl:
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/perl
+# Script to generate a 30x HTTP redirect (determined by the query parameter)
```
It might be better to use a python script since we are moving out of php/perl, see for instance LayoutTests/http/tests/xmlhttprequest/resources/redirect-cors.py

youennf commented on Source/WebCore/loader/cache/CachedResourceRequest.h:
```diff
@@ -106,6 +106,7 @@ class CachedResourceRequest {
     const SecurityOrigin* origin() const { return m_origin.get(); }
     SecurityOrigin* origin() { return m_origin.get(); }
 
+    bool containsFragmentIdentifier() const { return !m_fragmentIdentifier.isEmpty(); }
```
s/containsFragmentIdentifier/hasFragmentIdentifier/

stwrt replied:
done

youennf commented on LayoutTests/http/tests/links/resources/redirect-target.html:
```diff
@@ -0,0 +1,38 @@
+
+window.addEventListener("load", function () {
+    if (window.testRunner) {
+        shouldBe("window.location.hash", "#key=foo");
```
This line should be out of the `if(window.testRunner)`.

youennf commented on LayoutTests/http/tests/links/redirect-301-clear-memory-cache.html:
```diff
@@ -0,0 +1,17 @@
+    }
+    window.setTimeout(function() {
+      internals.clearMemoryCache();
+      window.location = 'resources/redirect-helper.pl?301#key=foo';
```
I am not sure to follow how this test is supposed to work.
The issue is that we reuse resources that were previously cached.
This test should first:
- cache a resource with #key=bar.
- load the same resource with #key=foo
- check we do not reuse the #key=bar resource.

I do not see the first step, I would expect this test to pass with and without the patch.
Can you check this?

stwrt replied:
This test case does work as we did also have a bug with when we had no fragments, but I'll add a test case to handle this scenario.

youennf commented on Source/WebCore/loader/cache/CachedResourceLoader.cpp:
```diff
@@ -1494,6 +1494,11 @@ CachedResourceLoader::RevalidationPolicy CachedResourceLoader::determineRevalida
     }
 
+    // FIXME: We should be able to USE the data here, but cannot currently due to a bug
+    // concerning redirect and URL fragments.
```
Can we file a bugzilla and add the bug URL to this FIXME.

stwrt replied:
Done. Will add here https://bugs.webkit.org/show_bug.cgi?id=258934

youennf commented on LayoutTests/http/tests/links/redirect-301-clear-memory-cache.html:
```diff
@@ -0,0 +1,16 @@
+        testRunner.waitUntilDone();
+    }
+    window.setTimeout(function() {
+      window.location = 'resources/redirect-helper-multi-redirect-clear-memory-cache.pl?301#key=foo';
```
From what I see the flow is:
- go to resources/redirect-helper-multi-redirect-clear-memory-cache.pl?301#key=foo
- this redirects to redirect-target-multi-redirect-clear-memory-cache.html#key=foo
- this navigates to redirect-helper-multi-redirect-final.pl?301#key=bar
- this redirects to redirect-target-multi-redirect-final.html#key=bar.

The 301 redirections are not the same so cache will not be hit.
I would expect this test to pass with and without the patch given this. Can you check this?

One approach that might be simpler to write and understand could be:
- Create an iframe that loads resources/redirect-helper.py?301#key=bar
- Once loaded, verify the iframe.contentWindow.location has '#key=bar'
- Optionally clear memory cache at this point (2 tests).
- Create another iframe that loads resources/redirect-helper.py?301#key=foo
- Once loaded, verify the iframe.contentWindow.location has '#key=foo'

It would also be good to add another test where the location header in the redirect contains #key=zip, in which case both frames should have the same iframe.contentWindow.location #key=zip.

stwrt replied:
Reverified the test cases without the patch and they do pass. Going to convert to the Iframe approach you recommended.

stwrt replied:
Fixed. Using `iframes` now.

I have 2 `iframes`. The first loads with the fragment. Then once the iframe loads `onload` is called and checks the `iframe` hash.

Then load the second fFrame. I then check the second hash is correct.

youennf commented on LayoutTests/http/tests/links/redirect-301-clear-memory-cache.html:
```diff
@@ -0,0 +1,44 @@
+
+            if (window.testRunner) {
+              testRunner.dumpAsText();
+              testRunner.notifyDone();
```
We could use testharness here, with promise_test without too much issues.
Ditto for the other tests I think.

youennf reviewed (approved):
r=me.
It would be good to modernise the tests: 
- use python (or even WPT python server) instead of perl 
- use testharness instead of custom test runner API.

stwrt committed:
URI Fragments Cached After 301 Redirect
https://bugs.webkit.org/show_bug.cgi?id=255924
rdar://problem/108535944

Reviewed by Youenn Fablet.

We end up reusing the fragments when performing a redirect and as such end up serving old parameters.
For now we will just disable caching redirects with fragments. This is a temporary fix; a more proper
solution of handling fragment caching in the future is required.

* LayoutTests/http/tests/links/redirect-301-clear-memory-cache-expected.txt: Added.
* LayoutTests/http/tests/links/redirect-301-clear-memory-cache.html: Added.
* LayoutTests/http/tests/links/redirect-301-no-inital-fragments-expected.txt: Added.
* LayoutTests/http/tests/links/redirect-301-no-inital-fragments.html: Added.
* LayoutTests/http/tests/links/redirect-301-with-fragments-expected.txt: Added.
* LayoutTests/http/tests/links/redirect-301-with-fragments.html: Added.
* LayoutTests/http/tests/links/resources/redirect-helper.pl: Added.
* LayoutTests/http/tests/links/resources/redirect-target.html: Added.
* Source/WebCore/loader/cache/CachedResource.cpp:
(WebCore::CachedResource::redirectReceived):
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::determineRevalidationPolicy const):
* Source/WebCore/loader/cache/CachedResourceRequest.h:
(WebCore::CachedResourceRequest::hasFragmentIdentifier const):
* Source/WebKit/NetworkProcess/cache/NetworkCache.cpp:
(WebKit::NetworkCache::makeUseDecision):
(WebKit::NetworkCache::makeStoreDecision):
* Source/WebKit/NetworkProcess/cache/NetworkCache.h:

Canonical link: https://commits.webkit.org/265906@main
